internal/aws/xray/telemetry/recorder.go

```diff
@@ -0,0 +1,384 @@
+)
+
+const (
+	envAWSHostname     = "AWS_HOSTNAME"
+	envAWSInstanceID   = "AWS_INSTANCE_ID"
+	metadataHostname   = "hostname"
+	metadataInstanceID = "instance-id"
+
+	queueSize       = 30
+	batchSize       = 10
+	defaultInterval = time.Minute
+)
+
+type Recorder interface {
+	// Start send loop.
+	Start()
+	// Stop send loop.
+	Stop()
+	// RecordSegmentsReceived adds the count to the record for the current period.
+	RecordSegmentsReceived(count int)
+	// RecordSegmentsSent adds the count to the record for the current period.
+	RecordSegmentsSent(count int)
+	// RecordSegmentsSpillover adds the count to the record for the current period.
+	RecordSegmentsSpillover(count int)
+	// RecordSegmentsRejected adds the count to the record for the current period.
+	RecordSegmentsRejected(count int)
+	// RecordConnectionError categorizes the error and increments the count by one for the current period.
+	RecordConnectionError(err error)
+}
+
+type RecorderOption interface {
+	apply(r *telemetryRecorder)
+}
+
+type recorderOptionFunc func(r *telemetryRecorder)
+
+func (ro recorderOptionFunc) apply(r *telemetryRecorder) {
+	ro(r)
+}
+
+func WithResourceARN(resourceARN string) RecorderOption {
+	return recorderOptionFunc(func(r *telemetryRecorder) {
+		r.resourceARN = resourceARN
+	})
+}
+
+func WithInstanceID(instanceID string) RecorderOption {
+	return recorderOptionFunc(func(r *telemetryRecorder) {
+		r.instanceID = instanceID
+	})
+}
+
+func WithHostname(hostname string) RecorderOption {
+	return recorderOptionFunc(func(r *telemetryRecorder) {
+		r.hostname = hostname
+	})
+}
+
+func WithLogger(logger *zap.Logger) RecorderOption {
+	return recorderOptionFunc(func(r *telemetryRecorder) {
+		r.logger = logger
+	})
+}
+
+type metadataProvider interface {
+	get() string
+}
+
+func getMetadata(providers ...metadataProvider) string {
+	var metadata string
+	for _, provider := range providers {
+		if metadata = provider.get(); metadata != "" {
+			break
+		}
+	}
+	return metadata
+}
+
+type simpleMetadataProvider struct {
+	metadata string
+}
+
+func (p simpleMetadataProvider) get() string {
+	return p.metadata
+}
+
+type envMetadataProvider struct {
+	envKey string
+}
+
+func (p envMetadataProvider) get() string {
+	return os.Getenv(p.envKey)
+}
+
+type ec2MetadataProvider struct {
+	client      *ec2metadata.EC2Metadata
+	metadataKey string
+}
+
+func (p ec2MetadataProvider) get() string {
+	var metadata string
+	if result, err := p.client.GetMetadata(p.metadataKey); err == nil {
+		metadata = result
+	}
+	return metadata
+}
+
+// ToRecorderOptions gets the metadata recorder options if enabled by the config.
+func ToRecorderOptions(cfg Config, sess *session.Session, settings *awsutil.AWSSessionSettings) []RecorderOption {
+	if !cfg.IncludeMetadata {
+		return nil
+	}
+	metadataClient := ec2metadata.New(sess)
+	return []RecorderOption{
+		WithHostname(getMetadata(
+			simpleMetadataProvider{cfg.Hostname},
```

Comment:
```suggestion
			simpleMetadataProvider{metadata: cfg.Hostname},
```

Please use field labels in struct initialization so that the structures can be evolved to include new fields if necessary.